src/librustc_codegen_ssa/mir/place.rs

```diff
@@ -388,7 +398,10 @@ impl<'a, 'tcx, V: CodegenObject> PlaceRef<'tcx, V> {
 
         // Cast to the appropriate variant struct type.
         let variant_ty = bx.cx().backend_type(downcast.layout);
-        downcast.llval = bx.pointercast(downcast.llval, bx.cx().type_ptr_to(variant_ty));
+        downcast.llval = bx.pointercast(
+            downcast.llval,
+            bx.cx().type_ptr_to(variant_ty, bx.cx().address_space_of_value(downcast.llval)),
```

Comment:
I was thinking of a good way to abstract it out, this is a good suggestion.

Although I've since removed the preservation aspect of this pull request, opting to pass the address space explicitly, and only in the few places it it actually required.